pkg/apiserver/rest/dualwriter_mode2_test.go

```diff
@@ -2,70 +2,297 @@ package rest
-	assert.NoError(t, err)
-	assert.Equal(t, 1, lsSpy.Counts("LegacyStorage.List"))
-	assert.Equal(t, 1, sSpy.Counts("Storage.List"))
-
-	resList, err := meta.ExtractList(l)
-	assert.NoError(t, err)
-
-	expectedItems := map[string]string{
-		// Item 1: Storage should override Legacy
-		"Item 1": "Storage field 1",
-		// Item 2 shouldn't be included because it's not in Storage
-		// Item 3 should because it's in Legacy
-		"Item 3": "Legacy field 3",
+var createFn = func(context.Context, runtime.Object) error { return nil }
+var exampleOption = &metainternalversion.ListOptions{
+	TypeMeta: metav1.TypeMeta{
+		APIVersion: "v1",
+		Kind:       "foo",
+	},
+}
+
+func TestMode2_Create(t *testing.T) {
+	type testCase struct {
+		name          string
+		input         runtime.Object
+		setupLegacyFn func(m *mock.Mock, input runtime.Object)
+		setupUSFn     func(m *mock.Mock, input runtime.Object)
+		wantErr       bool
 	}
+	tests :=
+		[]testCase{
+			{
+				name:  "creating an object in both the legacy and unified store",
+				input: exampleObj,
+				setupLegacyFn: func(m *mock.Mock, input runtime.Object) {
+					m.On("Create", context.Background(), input, mock.Anything, mock.Anything).Return(exampleObj, nil)
+				},
+				setupUSFn: func(m *mock.Mock, input runtime.Object) {
+					m.On("Create", context.Background(), input, mock.Anything, mock.Anything).Return(exampleObj, nil)
+				},
+			},
+			{
+				name:  "error when creating object in the legacy store fails",
+				input: failingObj,
+				setupLegacyFn: func(m *mock.Mock, input runtime.Object) {
+					m.On("Create", context.Background(), input, mock.Anything, mock.Anything).Return(nil, errors.New("error"))
+				},
+				wantErr: true,
+			},
+		}
+
+	for _, tt := range tests {
+		l := (LegacyStorage)(nil)
+		s := (Storage)(nil)
+		m := &mock.Mock{}
+
+		ls := legacyStoreMock{m, l}
+		us := unifiedStoreMock{m, s}
+
+		if tt.setupLegacyFn != nil {
+			tt.setupLegacyFn(m, tt.input)
+		}
+		if tt.setupUSFn != nil {
+			tt.setupUSFn(m, tt.input)
+		}
 
-	assert.Equal(t, len(expectedItems), len(resList))
+		dw := SelectDualWriter(Mode2, ls, us)
 
-	for _, obj := range resList {
-		v, ok := obj.(*dummyObject)
-		assert.True(t, ok)
-		accessor, err := meta.Accessor(v)
+		obj, err := dw.Create(context.Background(), tt.input, createFn, &metav1.CreateOptions{})
+
+		if tt.wantErr {
+			assert.Error(t, err)
+			continue
+		}
+
+		assert.Equal(t, obj, exampleObj)
+		accessor, err := meta.Accessor(obj)
 		assert.NoError(t, err)
+		assert.Equal(t, accessor.GetResourceVersion(), "")
+	}
+}
+
+func TestMode2_Get(t *testing.T) {
+	type testCase struct {
+		name          string
+		input         string
+		setupLegacyFn func(m *mock.Mock, input string)
+		setupUSFn     func(m *mock.Mock, input string)
+		wantErr       bool
+	}
+	tests :=
+		[]testCase{
+			{
+				name:  "getting an object from storage",
+				input: "foo",
+				setupLegacyFn: func(m *mock.Mock, input string) {
```

Comment:
Could be helpful to turn this into a test case where we have both functions successfully return an object and no errors. In addition, we could have Storage return a slightly different version of the same object. For example, the name is the same as the object returned by LegacyStorage, but it additionally has labels. That way it makes it easier to demonstrated that the object being fetched comes from Storage even though it also exists in LegacyStorage.

Reply:
I added this case in a different test and not here, for some reason. Just added the test :)